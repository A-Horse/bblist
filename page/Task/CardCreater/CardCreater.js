import React, { Component } from 'react';
import { createTaskCard } from 'actions/task/task-card';
import { getTaskAllCards } from 'actions/task/task-wall';
import { addBodyEventListenerOnce } from 'actions/event/body';
import { Button } from 'components/widget/Button/Button';
import { Hr } from 'components/widget/Hr';
import UserAvatar from 'components/UserAvatar';
import { MoreIcon } from 'services/svg-icons';
import { IconAdd } from 'services/image-icon';

// import 'style/page/task/taskcard-creater.scss';
import './CardCreater.scss';

class CardCreater extends Component {
  constructor(props) {
    super(props);
    this.close = this.close.bind(this);
    this.toggle = this.toggle.bind(this);
    this.createCard = this.createCard.bind(this);
  }

  componentWillMount() {
    this.state = {
      toggle: false
    };
  }

  clearInput() {
    this.refs.taskCardTitle.value = '';
  }

  createCard() {
    const data = {
      taskBoardId: +this.props.wallId,
      taskListId: +this.props.listId,
      title: this.refs.taskCardTitle.value.trim()
    };
    return dispatch(createTaskCard(data)).then(() => {
      this.clearInput();
      dispatch(getTaskAllCards(this.props.wallId)).then(this.close.bind(this));
    });
  }

  close() {
    this.setState({ toggle: false });
  }

  toggle(event) {
    this.setState({ toggle: true });
  }

  render() {
    if (this.state.toggle) return this.renderBody();
    return this.renderToggle();
  }

  renderToggle() {
    return (
      <div onClick={this.toggle} className="taskcard-creater--toggle">
        <IconAdd className="icon-add" />
        <span>Add a card...</span>
      </div>
    );
  }

  renderBody() {
    return (
      <div className="taskcard-creater--body" onClick={event => event.stopPropagation()}>
        <div>
          <textarea
            type="text"
            ref="taskCardTitle"
            placeholder="Task Content"
            onKeyPress={event => {
              if (event.ctrlKey && event.key === 'Enter') this.createCard.bind(this)();
            }}
          />
        </div>
        <div className="taskcard-creater--user">
          <UserAvatar user={this.props.loginedUser.toJS()} />
          <span>
            {this.props.loginedUser.get('username')}
          </span>
        </div>
        <Hr />
        <div className="taskcard-creater--participants">
          <div>Participants</div>
          <div>
            <UserAvatar user={this.props.loginedUser.toJS()} />
          </div>
        </div>
        <Hr />
        <div className="taskcard-creater--operation">
          <div>
            <MoreIcon />
            <span>more</span>
          </div>
          <div>
            <Button className="btn-cancel" styleType="default" onClick={this.close}>
              Cancel
            </Button>
            <Button styleType="primary" onClick={this.createCard}>
              OK
            </Button>
          </div>
        </div>
      </div>
    );
  }
}

export default CardCreater;
/* const mapStateToProps = state => {
 *   return {
 *     user: state.auth.loginedUser
 *   };
 * };
 *
 * export default connect(mapStateToProps)(TaskCardCreater);*/
