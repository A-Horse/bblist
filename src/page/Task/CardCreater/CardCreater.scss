@import "style/variable";
$list-horiz-padding: 10px;
$list-vertical-padding: 12px;

.taskcard-creater--body {
  background-color: $content-light;
  box-shadow: 0 2px 4px $gray-light-litte-depth;
  margin-top: 4px;
  padding: $list-horiz-padding $list-vertical-padding;
  transition: opacity 300ms ease-in-out;

  &.before-close {
    opacity: 0.1;
  }

  textarea {
    width: 100%;
    border: 1px solid $border-light;
    border-radius: $radius-middle;
    box-shadow: 0 1px 3px $gray-light-litte-depth inset;
    font-size: $font-size-x-middle;
    padding: 8px 11px;
    box-sizing: border-box;
  }
}

.taskcard-creater--toggle {
  font-weight: bolder;
  color: $gray-deep;
  margin: 8px 10px 10px;
  cursor: pointer;

  span {
    vertical-align: middle;
  }

  .fa-plus {
    vertical-align: middle;
    margin-right: 8px;
  }
}

.taskcard-creater--user,
.taskcard-creater--operation,
.taskcard-creater--participants {
  margin-top: 10px;
  font-size: $font-size-x-middle;
  color: $gray-light-deep;
}

.taskcard-creater--user,
.taskcard-creater--participants {
  margin-bottom: 8px;
}

.taskcard-creater--user {
  display: flex;
  align-items: center;

  span {
    margin-left: 12px;
  }
}

.taskcard-creater--operation {
  .btn-cancel {
    margin-right: 20px;
  }
}

.taskcard-creater--operation,
.taskcard-creater--operation > div {
  display: flex;
  justify-content: space-between;
  align-items: center;
}

.taskcard-creater--participants {
}
