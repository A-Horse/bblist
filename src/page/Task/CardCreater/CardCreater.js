// @flow
import React, { Component } from 'react';
import PropTypes from 'prop-types';
import { Button } from '../../../components/widget/Button/Button';
import { UserAvatar } from '../../../components/UserAvatar/UserAvatar';
import ClickOutSide from '../../../components/utils/ClickOutSide';
import { timeout } from '../../../utils/timeout';

import { Select } from 'antd';

const Option = Select.Option;

import './CardCreater.scss';

class CardCreater extends Component<
  {
    loginedUser: PropTypes.object.isRequired,
    addTaskCard: PropTypes.func.isRequired,
    track: PropTypes.object.isRequired
  },
  {
    title: '',
    toggle: false,
    beforeClose: false,
    type: 'STORY'
  }
> {
  state = {
    title: '',
    toggle: false,
    beforeClose: false,
    type: 'STORY'
  };

  clearInput() {
    this.taskCardTitle.value = '';
  }

  addCard = () => {
    this.props.addTaskCard({ title: this.state.title.trim(), type: this.state.type });
    this.clearInput();
    this.close();
  };

  close = async () => {
    this.setState({ beforeClose: true });
    await timeout(200);
    this.setState({ toggle: false, beforeClose: false });
  };

  toggle = async () => {
    this.setState({ toggle: true });
    await timeout();
    window.document.querySelector(
      `[data-id="${this.props.track.get('id')}"].task-track .task-track--body`
    ).scrollTop = 10000;
  };

  handleTaskTypeChange = value => {
    this.setState({ type: value });
  };

  render() {
    if (this.state.toggle)
      return (
        <ClickOutSide onClickOutside={this.close}>
          <div
            className={`taskcard-creater--body${this.state.beforeClose ? ' before-close' : ''}`}
            onClick={event => event.stopPropagation()}
          >
            <div>
              <textarea
                autoFocus
                type="text"
                ref={ref => (this.taskCardTitle = ref)}
                placeholder="Task Content"
                onChange={event => this.setState({ title: event.target.value })}
                onKeyPress={event => {
                  if (event.ctrlKey && event.key === 'Enter') this.addCard();
                }}
              />
            </div>

            <div>
              <Select
                defaultValue="STORY"
                style={{ width: 120 }}
                onChange={this.handleTaskTypeChange}
              >
                <Option value="STORY">Story</Option>
                <Option value="TODO">Todo</Option>
              </Select>
            </div>

            <div className="taskcard-creater--user">
              <UserAvatar user={this.props.loginedUser.toJS()} />
              <span>{this.props.loginedUser.get('username')}</span>
            </div>
            <div className="taskcard-creater--participants">
              <div>Participants</div>
              <div>
                <UserAvatar user={this.props.loginedUser.toJS()} />
              </div>
            </div>
            <div className="taskcard-creater--operation">
              <div>
                <i className="fa fa-ellipsis-h" aria-hidden="true" />
                <span>more</span>
              </div>
              <div>
                <Button className="btn-cancel" styleType="default" onClick={this.close}>
                  Cancel
                </Button>
                <Button styleType="primary" onClick={this.addCard}>
                  OK
                </Button>
              </div>
            </div>
          </div>
        </ClickOutSide>
      );
    return (
      <div onClick={this.toggle} className="taskcard-creater--toggle">
        <i className="fa fa-plus" aria-hidden="true" />
        <span>Add a card...</span>
      </div>
    );
  }
}

export default CardCreater;
