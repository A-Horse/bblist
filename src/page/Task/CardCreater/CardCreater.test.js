import React from "react";
import CardCreater from "./CardCreater";
import renderer from "react-test-renderer";

test("Task CardCreater", () => {
  const tree = renderer.create(<CardCreater />);
  expect(tree).toMatchSnapshot();
});
