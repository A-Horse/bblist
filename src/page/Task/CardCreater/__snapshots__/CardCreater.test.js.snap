// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Task CardCreater 1`] = `
<div
  className="taskcard-creater--toggle"
  onClick={[Function]}
>
  <i
    aria-hidden="true"
    className="fa fa-plus"
  />
  <span>
    Add a card...
  </span>
</div>
`;
